Pattern mvvm

VitaliiDiadchenko opened the pull request:
Added pattern MVVM

VitaliiDiadchenko committed:
Added dependencies lifecycle viewmodel and livedata
Created viewModel package
Created ViewModelMovieListFactory.kt
Created ViewModelMovieList.kt
Added progressBar to fragment_movies_list.xml
Added viewModel to FragmentMovieList.kt

VitaliiDiadchenko committed:
Created ViewModelMovieDetail.kt
Created ViewModelMovieDetailFactory.kt

VitaliiDiadchenko committed:
Fix mistakes in ViewModelMovieDetail.kt
Fix height size imagePosterCard

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/viewModel/ViewModelMovieDetailFactory.kt:
```diff
@@ -0,0 +1,14 @@
+package com.vitaliidiadchenko.mykotkinapplication.viewModel
```
Для навигации по коду приложения будет удобнее сложить в один пакет все классы, которые относятся к одному экрану: фрагмент, вью модель, вью модель фабрику, адаптеры для ресайклера и т.д.

Идея в том, чтобы все относящееся к одному классу лежало в одной кучке)

VitaliiDiadchenko replied:
Да, хотел так сделать, а потом подумал что лучше спрошу как правильней. Т.е. получается и адаптеры тоже в один пакет с фрагментами?

an-yurii replied:
Да. Все, что относится к конкретному фрагменту, лучше положить с ним рядом.

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/viewModel/ViewModelMovieDetailFactory.kt:
```diff
@@ -0,0 +1,14 @@
+import com.vitaliidiadchenko.mykotkinapplication.data.Movie
+import java.lang.IllegalArgumentException
+
+class ViewModelMovieDetailFactory (private val movie: Movie): ViewModelProvider.Factory {
```
Экран с отображением фильма - статический. Нет какой-либо надобности использовать вью модель.

VitaliiDiadchenko replied:
Сбило с толку что для каждого фрагмента своя вью модель.

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/FragmentMovieDetail.kt:
```diff
@@ -45,8 +48,9 @@ class FragmentMovieDetail : Fragment() {
         arguments?.getParcelable<Movie>("movie")?.let { movie ->
-            setupView(movie)
-            Log.i("movie", movie.id.toString())
+            val viewModelFactory = ViewModelMovieDetailFactory(movie)
```
Это - статический фрагмент, который отображает информацию о фильме. В этот момент у есть вся информация о фильме, ее нужно просто показать.
Этот экран не нужно было менять)

VitaliiDiadchenko replied:
Исправлю.

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/FragmentMovieList.kt:
```diff
@@ -5,22 +5,24 @@ import android.os.Bundle
+    private var progressBar: ProgressBar? = null
     private var listener: FragmentListener? = null
-    private val scope = CoroutineScope(Dispatchers.IO)
+    private lateinit var viewModel: ViewModelMovieList
```
Обрати внимание, как сделана инициализация в воркшопе
```
private val viewModel: Workshop2SolutionViewModel by viewModels { Workshop2SolutionViewModelFactory() }
```

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/viewModel/ViewModelMovieList.kt:
```diff
@@ -0,0 +1,40 @@
+import kotlinx.coroutines.launch
+import java.lang.Exception
+
+class ViewModelMovieList(
```
Лучше использовать постфиксную нотацию `MovieListViewModel`. В этом случае файлы будут группироваться по именам, удобнее навигироваться в проекте.

an-yurii commented on app/src/main/java/com/vitaliidiadchenko/mykotkinapplication/viewModel/ViewModelMovieListFactory.kt:
```diff
@@ -0,0 +1,23 @@
+import kotlinx.coroutines.Dispatchers
+
+
+class ViewModelMovieListFactory(private val context: Context,
```
Лучше использовать постфиксную нотацию `MovieListViewModelFactory`.
- MovieListViewModel
- MovieListViewModelFactory

an-yurii reviewed (approved):
Хорошая работа)

VitaliiDiadchenko committed:
Removed DetailViewModel
Create class App
Added dependencies fragment
Create package screens
Refactor files to packages
Rename files

VitaliiDiadchenko committed:
Refactor files from adapter package
Remove adapter package
